### ISSUE TITLE
Filtering for un-revoked objects does not work without explicit revoked=False

### ISSUE THREAD

**Issue by oliveriandrea** [oasis-open/cti-python-stix2#123]
I use Python 3.6, with the latest github stix2 version
If you try to filtre for revoked (or not revoked) objects on FileSystemStore and MemoryStore you obtain errors or incorrect behaviours

Example
`from stix2 import FileSystemStore, MemoryStore, Filter, AttackPattern`
`fs = FileSystemStore('/Users/andrea/test')`
`mem = MemoryStore()`
`filt_false = Filter('revoked', '=', False)`
`filt_true = Filter('revoked', '=', True)`
`patt1 = AttackPattern(name='test1')`
`patt2 = AttackPattern(name='test2')`
`fs.add(patt1)`
`mem.add(patt2)`

`fs.query(filt_false)`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/__init__.py", line 90, in query
    return self.source.query(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/filesystem.py", line 298, in query
    stix_obj = json.load(open(os.path.join(root, file_)))
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/__init__.py", line 296, in load
    return loads(fp.read(),
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 1083: invalid start byte

`fs.query(filt_true)`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/__init__.py", line 90, in query
    return self.source.query(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/filesystem.py", line 298, in query
    stix_obj = json.load(open(os.path.join(root, file_)))
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/__init__.py", line 296, in load
    return loads(fp.read(),
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 1083: invalid start byte

`mem.query(filt_false)`
[]

`mem.query(filt_true)`
[]

`patt1 = patt1.revoke()`
`patt2 = patt2.revoke()`
`fs.add(patt1)`
`mem.add(patt2)`

`fs.query(filt_true)`
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/__init__.py", line 90, in query
    return self.source.query(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/stix2/sources/filesystem.py", line 298, in query
    stix_obj = json.load(open(os.path.join(root, file_)))
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/json/__init__.py", line 296, in load
    return loads(fp.read(),
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 1083: invalid start byte

`mem.query(filt_false)`
[]

`mem.query(filt_true)`
[AttackPattern(type='attack-pattern', id='attack-pattern--3e65f786-25a0-471d-8ab8-1929313424a3', created='2018-01-19T09:29:36.104Z', modified='2018-01-19T09:30:49.983Z', name='test2', revoked=True)]

**Comment by gtback** [comment]
Hmmm, that's strange. Thanks for reporting, @oliveriandrea . We'll have to look into it.

**Comment by oliveriandrea** [comment]
Do you have some news about it?

**Comment by gtback** [comment]
Took a quick look. There seem to be a couple issues involved:

- The UnicodeDecodeError is caused by files in that directory that aren't valid JSON (like .DS_Store on macOS).
- Even after I fixed that, I'm still not getting the data I expect returned, so there's something else going on that I'll have to look into more.

**Comment by mbastian1135** [comment]
1. Decode errors. These should be fixed (or mitigated as best can), as now when the FileSystemStore/Source combs through the directory, it ignores non ".json" files, raises error if bad JSON, or raises error if bad STIX form. ( This patch is in Master branch as of 21-March-2018)
2.  Query errors. I cannot reproduce them. @oliveriandrea , I reproduced your steps and got expected results. Maybe try again, with latest Master branch. Apologies for delay, we were off the project for a while. @gtback , do you remember what issues you were noticing.

**Comment by gtback** [comment]
I just tried out again with the latest master. The decode errors are gone, but I realized a significant limitation of the current filter implementation. Even though `false` is the default value for `revoked`, unless you explicitly set `AttackPattern(..., revoked=False)`, it won't get returned by a query. I'm not sure the best way to fix this. Any ideas, @mbastian1135 ?

**Comment by mbastian1135** [comment]
Right, so the filtering logic does this because "revoke" is not found as a property on the object (as its optional), and thus returns False because the filter cannot be applied. 

If we reversed this logic, I think we would cause way more chaos than its worth.  Since then the filtering mechanism would return objects that are completely not inline with the filter (e.g. imagine querying for labels="Trojan" and any object that does not have the labels property would be returned)

An alternative is a hack within the filtering mechanism that checks if the filter being applied is evaluating a boolean property and that boolean property is True by default (as specified by STIX spec). In this case the STIX object would circumvent the normal disregarding and be returned. Id put that check here: https://github.com/oasis-open/cti-python-stix2/blob/master/stix2/datastore/filters.py#L141

**Comment by chisholm** [comment]
> If we reversed this logic, I think we would cause way more chaos than its worth. Since then the filtering mechanism would return objects that are completely not inline with the filter (e.g. imagine querying for labels="Trojan" and any object that does not have the labels property would be returned)

Unless the `labels` property has a default value, I don't think this is true.  Greg said that `false` is the default for the `revoked` property.  This means the property *always has a value*, though it may not have been explicitly given.  The "labels" property doesn't have a default (does it?), so it's not the same thing.

It sounds like the filter code cares whether the property value was explicitly given.  And that sounds wrong.  For most uses, it don't think it matters.  One case where it would, is during serialization, if you want to minimize verbosity (by omitting the property) when a value isn't explicitly given, and exactly reproduce the input on output when it is explicitly given.  But these are cosmetic issues, so from a correctness standpoint, they don't matter either.

The AttackPattern SDO looks like its revoked property was written as:
```python
('revoked', BooleanProperty())
```
which doesn't give a default.  If it defaults to false, shouldn't a default be given here?  E.g.
```python
('revoked', BooleanProperty(default=False))
```
Or am I misunderstanding?  (Maybe, it's been awhile since I've looked at the spec.)

**Comment by clenk** [comment]
We implemented property defaults, we [assumed ](https://github.com/oasis-open/cti-python-stix2/blob/ed2cf28cdf17cde2aff2ecce19274b6d8fc9299d/stix2/properties.py#L57) having a default means the property is required, too. So that's why we didn't use `BooleanProperty(default=False)` in this case. I believe so far the only places we've used defaults are on timestamp properties that are also required. The `revoked` common property and the `summary` property on `Sighting` objects are the only examples I found of the spec listing a default value for any property, and both are optional properties.

@chisholm's suggestion is simple, but will it confuse users if the property shows up when they didn't specify it? A round-trip of parsing some JSON into python-stix2 and then printing it back out to JSON won't result in the original input. They will be semantically the same though.

Maybe we need to decouple our assumption that having a default = required. We could modify [this line](https://github.com/oasis-open/cti-python-stix2/blob/c98abb18f54e2f0fd1dae2df8c20339ef5387279/stix2/base.py#L54) to also check if the property is required. Then the filter code would check if the object type has any non-required properties with defaults. This would keep us from maintaining a separate list of "optional-with-defaults" properties.

**Comment by chisholm** [comment]
Boy, there's definitely terminology confusion here :)  At the conceptual level, a defaulted property is a required property.  I think you meant in your first sentence to say that defaulted in the current API design implied *usage* of the property in the API would be required (when constructing an SDO).  And you didn't want usage of `revoked` to be required, so you didn't default it either.  This is really mixed up thinking, to me.

I actually wrote about this before; I just dug up the email, which is dated 3/27/2017, almost exactly one year ago.  But was only addressed to Greg at the time.  It is about stix2 API design, specifically about the `Property` constructor.  Here's a paste:

>to my mind, fixed implies defaulted and required, and default implies required.  So they're all dependent.  So instead of making three independent choices which may not make sense together, just choose one of the three.

This is a statement about the fixed/defaulted/required concepts, which I hoped to use to drive API design (I didn't like the `Property` constructor having those three separate kwargs).  Yes, defaulted implies required, but it doesn't imply *usage* of the property is required, when constructing an SDO.  The whole point of having a default is so that you don't have to specify the property: when omitted, its value is set to the default.  You default a Python function argument, for example, specifically so that callers *don't* have to specify it, yet you still maintain the invariant that the argument always has a value.  The function requires a value for the argument (therefore it is effectively required), but you don't want callers to always have to specify one.  You get the nice invariant, without burdening users with overly verbose function calls.  Defaulting is a convenience.

In the STIX spec, when a property is defaulted, that means it is conceptually required.  I know the spec has "(optional)" after the `revoked` property name, but that's only about usage.  When you default a property, it becomes impossible for it to not have a value.  If you were to implement a "delete" method which could operate on a defaulted property, you'd have to implement it with behavior like "reinstate the default".  You can't really remove the property, since by definition (of "default"), if not otherwise specified, the property has its default value.  (Which means it must exist.)

You wouldn't default a property unless you were thinking in your mind that it always had to be present with a value.  That it wouldn't make any sense for the property to not exist.  I.e. that it is conceptually required.  You only want usage to be optional.  So if the current API design is that giving a default value means that the property must always be given an explicit value when the SDO is constructed, then the API is designed wrong.  I would think that runs counter to most programmers' intuitions and experience with defaulting.

The design I wanted was for a `Property` creator to make a single choice, and for the API to be clearer about this:
 * **optional**: the property is optional.  If not given, it will not exist on the SDO.  This could be the... default :)
 * **required**: the property is required, and has no default.  This implies that *usage* is required.  It is the only way the property can obtain a value.  (Note, this is a slightly stricter definition from what I've been talking about conceptually, in that there's the additional "and has no default" part.  But I think users will get it.  Can you think of a better name?)
 * **defaulted**: the property is required, but has a default.  This implies that *usage* is optional.  If not explicitly given, the property's value will be the default.
 * **fixed**: the property is defaulted to a given value, and additionally can't be overridden by giving an alternative value explicitly during SDO construction.  You may explicitly give the same value as the default, but that would have exactly the same effect as if the property were omitted.

The way the API looked to me, with the three keywords, is that you can pick any combination of the three.  That's the way keyword args usually work.  It didn't seem to fit the conceptual requirements of the property.  That was my take on it, anyway.

As far as "remembering" whether a defaulted/fixed parameter were explicitly given or not, you wouldn't have to worry about properties mysteriously showing up.  You'd have to worry about them disappearing.  This could happen if a defaulted/fixed property were explicitly assigned the default/fixed value.  During serialization, the library could notice that and omit it, knowing the semantics would be the same.  If you didn't want that, extra bookkeeping would be required.  You'd have to keep track of which defaulted/fixed properties were explicitly assigned their default/fixed values, with the interpretation that "now the user wants these to be explicit in their JSON serializations".  I wouldn't assume that to be true though, I would prefer a user configurable setting.

**Comment by gtback** [comment]
We originally used "default" in the sense of "this field is required, but this is how you can come up with a sensible value if one isn't provided". As @clenk  pointed out, the only place we've used this is for created/modified timestamps. Perhaps we can rename that to something like "auto_now" (taking inspiration from Django's DateTimeFields. I like the idea of separating this out from a "default" value (from the spec), which semantically exists even if not present in the JSON. 

It would be a lot of bookkeeping if we want to track specific default fields on specific instances of objects to know whether to export a value of `"revoked": false`, but I agree we can compromise on using either a global setting or add it to the serialize() call.

**Comment by chisholm** [comment]
> I like the idea of separating this out from a "default" value (from the spec), which semantically exists even if not present in the JSON.

I think you can use the same mechanism.  In fact, two different mechanisms that both do "defaults" in slightly different ways might be too confusing.  If the spec mandates a default and the user didn't supply a value, the library should automatically supply the spec's default.  Otherwise, to my mind, the library isn't being compliant.

Your additional defaulting behavior sounds just like what the higher convenience-layer factories were envisioned to do, I think.  Do we need it in both places?

> It would be a lot of bookkeeping if we want to track specific default fields on specific instances of objects to know whether to export a value of "revoked": false, but I agree we can compromise on using either a global setting or add it to the serialize() call.

I didn't mean to advocate it, I was just addressing @clenk 's concern regarding unexpected property behavior.  Ideally, we could have a simple, natural implementation, and not have to worry about users being surprised by serialization details.

**Comment by mbastian1135** [comment]
FTR: after dev meeting (2018 March 28) there was general consensus about:
- modifying the creation of properties to specific mutually exclusive options (i.e. optional, required, defaulted etc..)  for a more clearer, expressive creation pf properties(long-term, especially before 1.0 release)
- Not doing bookkeeping for tracking of properties that will/will not be serialized/exported but rather allowing a user to specify at some level whether to export "semantically default" properties
- Further exploration of how many types of STIX properties fit into these corner cases of being optional (from a STIX specification perspective) but semantically defaulted from an API perspective. As Chris noted, he found "revoked" and "Sighting.summary" to be the only instances of optional STIX properties that also had defaults.

(Note Im paraphrasing here and enumerating from memory. So @gtback , @clenk please correct any incorrect statements)

**Comment by gtback** [comment]
I looked and didn't find any other default properties. I also noticed that both are required properties, so maybe we can handle this by making it an attribute of the `BooleanProperty` class? The only other required `boolean` property is `EmailMessage.is_multipart`, which doesn't have a default value, so it is `required` in the sense that @chisholm describes above (user must provide a value).

**Comment by gtback** [comment]
I think what @mbastian1135 said above (primarily what @chisholm suggested) makes the most sense:

1. Remove the 'fixed', 'required', and 'default' kwargs for `Property` (sub-)classes. 
2. Create a new kwarg `occurrence` (I don't love the name, but it's the best I can think of right now) that has one of four values:
    - `optional` (the default). Properties of this type don't require the user to provide a value, and if they don't, nothing will show up in the serialized output. For now, there's no way to filter on objects where these values are not set (but maybe we can extend the filter syntax later).
    - `required`. Properties of this type are required by the spec, and have no value that can be inferred by the library. Users **MUST** provide a valid value when constructing the object
    - `fixed`. Properties of this type are required by the spec to have a specific value. Users **MAY** provide this fixed value when constructing the object, but regardless of whether they do or not, it will appear in the serialized output.
    - `defaulted`. Properties of this type are required (semantically) by the spec, but the library can help users fill them in. I think there are only four of these: `created` and `modified` time (which can be set to the current time), and `revoked` and `Sighting.summary` (Boolean properties that default to False). These will appear in the serialized output (even in the default False case).  No need to do bookkeeping on whether these were actually provided.
3. Add a `value` kwarg that is used for `fixed` and `defaulted` Properties. A special sentinel value (`NOW`) can be used for the default timestamp properties. 

Actually, now that I think about it, maybe "Fixed" and "Defaulted" should be property classes rather than special `occurrence` values. I haven't thought through this yet, but it's late on a Friday, so I'll take the weekend to consider it more.

**Comment by chisholm** [comment]
I'm okay with invocations looking like:
```python
SomeProp(required=True)
SomeProp(default="foo")
SomeProp(fixed="foo")
```

I think it's at least as readable as:
```python
SomeProp(occurrence="required")
SomeProp(occurrence="defaulted", value="foo")
SomeProp(occurrence="fixed", value="foo")
```

I assume you intend to enforce legal values on the `occurrence` parameter.  It's just as easy to enforce legal keys in a keyword arg, and you can get something a little more streamlined.  But either way, you should do that.  (Hopefully this doesn't seem like some kind of backpedaling, since that's kinda how the current API already is.  I don't know if this is seeming unclear or going around in circles.  I don't think so, but maybe I'm making people hopelessly confused :-P)

**Comment by gtback** [comment]
@chisholm, are you saying you're OK with having different kwargs for `required`, `default`, and `fixed` as long as the library throws an error if you try to use more than one of them, and that this is properly documented? It sounded like you thought there should only be one kwarg for the "occurrence type" of the property, but maybe I misunderstood.

**Comment by chisholm** [comment]
Yeah, that's what I was afraid of.  My (probably too long) comment was really about the concepts, since I felt there was confusion.  All the code and design flows from that, so I wanted to clarify the ideas.  And it was the second time it had come up (for me, anyway), so it seemed like a recurring problem.  Also, I felt like library was not being spec-compliant (regarding defaults), which is a more serious issue.  I think we agree about the concepts; this remaining confusion is about API design.  Apologies if I seemed to nitpick over abstract concepts. That original email was written much earlier in the API development, so I think it was more appropriate to critique the design then.  Maybe that ship has sailed.

But in case it matters or would still help, here are some concrete ideas about the API design.  So, here's the current `Property` constructor:

```python
def __init__(self, required=False, fixed=None, default=None, type=None):
    # etc ...
```

All parameters are defaulted.  From our above discussion, that means they're all required.  But `Property` conceptually needs a single choice.  All required... yet you must choose exactly one.  Can you see the conflict?  Design-wise, that was the main issue I had.  You might think "yeah, but they default to None", but to me, it doesn't change much.  They're always all there with a value, though it may sometimes be None.  And yeah, there is no error checking and not enough documentation.  What's basically been set up is a function with a set of defaulted parameters with a bunch of co-constraints among them (e.g. if arg X is *this*, then arg Y must be *that*).  In general, I think that makes it harder to understand, and more apt to be used incorrectly.

So here how I designed it:

```python
def __init__(self, type_, **kwargs):
    if len(kwargs) > 1:
        raise AttributeError("Too many usage args for property "
                             "({} > 1)".format(len(kwargs)))

    for kw in kwargs:
        if kw not in ("fixed", "default", "required"):
            raise AttributeError("Unrecognized keyword arg: {}".format(kw))

    # etc...
```
Here, there are no defaults.  It's generic `**kwargs`.  That in itself gives no guidance as to what to pass, but I wrote a big docstring (omitted), and the very first bits of code tell you exactly what's expected: there can be no more than one keyword arg, and if a keyword arg is given, it must be one of those three values.  The language can't exactly express the requirements in the formal parameter list, but at least it doesn't mislead.

Clear as mud now?? :)  These kinds of discussions are great during design time, but perhaps a moot point now.  Sorry for any heartburn!

**Comment by chisholm** [comment]
>  But checking the values in a **kwargs parameter against a fixed list of acceptable values smells like an anti-pattern to me.

I try to achieve a design which is closest to the model in my head.  No language allows you to express every idea perfectly.  In this case, the ideal is to express that at most one keyword arg is required, where the keyword is one of a fixed set.  To me, listing all three as defaulted keyword args is expressing the wrong thing.  I don't know if it's considered some over-arching language anti-pattern, but in this situation, it feels like the right thing to do.  Well, the best of the available options, anyway.

How would you express the at-most-one constraint with three None-defaulted parameters?
```
if (fixed and required) or (fixed and defaulted) or ...
    # raise error
```
Yuck.  That doesn't scale well (with the number of keyword args).  Maybe counting the `None`s?
```python
if sum(1 if a is None else 0 for a in (required, fixed, default)) < 2:
    # raise error
```
Scales better, but still awkward.  Or maybe you'd prefer:
```python
if fixed:
    if default or required:
        # error
    # handle fixed
elif default:
    if fixed or required:
        # error
    # handle default
# etc...
```
That's pretty verbose.  You can probably think of other ways.  Can you think of any good understandable ones?  Maybe this is why the check was never done.  There didn't seem to be a simple natural way to do it.  Maybe it's because that function signature just doesn't really fit the concept very well :)  Or maybe I'm just missing the obvious.

If you smell an anti-pattern, maybe the usage/value split parameters would be less smelly to you?  And bypass the need for an at-most-one-usage check.

>  I see the benefits of moving type_ up to the first argument.

I should clarify that was pasted from code whose design diverged in more ways than just the `Property` constructor.  My `type_` arg is a completely different animal.  It is more like the various property subclasses.  I wanted to experiment with splitting the typing from the properties.  So my design only needed one `Property` class, but there are many types (subclasses of a `_STIXType` class).  Anyway... not important for this.  The main point was the `**kwargs` style, with the error checking.  You can ignore the `type_` arg.  Obviously I had to move it up to the front since I made it positional.

> When I'm talking about implicit semantic default values of properties, I mean values of Propertys in STIXBase objects, not the value of keyword arguments to Property.\_\_init\_\_.

Of course.  One rarely thinks of functions as having "properties".  But I don't think there's one definition of "default" which applies to SDO properties, and another to Python functions.  I was applying the same general concept to both.  I thought it fit in both contexts.

> The way that Propertys are defined in python-stix2 is *almost* an internal implementation detail, except that's also how users can define Custom Objects.

Sure, but one can still have aesthetics and care about design.  This conversation probably should have happened a year ago... :)

**Comment by mbastian1135** [comment]
FTR - Dev Meeting 11 April 2018:
- Proposed simple solution of having the two properties (revoked, sightings.summary) have default values of "False". This solves the usage bug that this thread was originally started on. But need to think through the effects of doing this for serialize() and python-stix2 output in general.
- The above solution is to address the issue in the short term before 1.0 release, the deeper issue of the design of the Property object is still open.

(@clenk, @gtback - correct if I misinterpreted)

**Comment by gtback** [comment]
This was fixed for 1.0. I'll open a separate issue to consider refactoring how Properties are designed.
